### PR TITLE
Add support for express accounts

### LINKED ISSUE #5058: PayPal Commerce should support Express Checkout acounts
## User Story

As a citizen of a country that can't use custom payments with PayPal, I want to be able to use Smart Buttons (Express Checkout) so that I can still process payments with PayPal Commerce.

## Details
Only a fixed list of countries within the world can currently use custom payments (PPCP_CUSTOM), which means that everyone else can only use Smart Buttons (EXPRESS_CHECKOUT). We need to add support for Express Checkout to the onboarding process as well as the payment form.

### Onboarding

We will pass the two-digit country code via the additional `country_code` POST parameter to the PayPal link gateway call. The gateway will then determine whether the account can be `PPCP_CUSTOM` or `EXPRESS_CHECKOUT`.

When the user finishes onboarding and the merchant details are sent to the server, we only need to check that their email is verified and payments — we don't need to check the product or capabilities. We then need to mark the `MerchantDetail` as being able to process custom payments or not.

### Donation Form

The donation form needs to change depending on whether the account can process custom payments or not. If the account _can_, then it works exactly as it does now with the combination of Smart Buttons and Advanced Fields. If the account _cannot_ process custom payments, then do not load the advanced fields. Instead, display the Smart Buttons and include the "Debit or Credit Card" button.

## Acceptance Criteria

-   [x] Country is passed to the gateway for the Partner link
-   [x] Only the `payments_receivable` and `primary_email_confirmed` are check in onboarding
-   [x] The account is marked as being able to make custom payments based on whether its EXPRESS or PPCP
-   [x] The Advanced fields do not show up for Express accounts
-   [x] The Smart Buttons work for Express accounts
-   [x] The "Debit or Credit Card" Smart Button _only_ shows up for Express accounts

### PR DESCRIPTION
**Pull request by JasonTheAdams**
Resolves #5058 

## Description

This PR wraps up support for Express accounts. By default, accounts will attempt to be PPCP, which supports custom payments (i.e. advanced fields), but only 28 countries support PPCP. All other countries will need to use Express, which limits them to Smart Buttons.

Incidentally, PayPal itself actually checks for the existence of Advanced Fields. If they're not present, then the "Debit or Credit Card" button shows up in the Smart Buttons, so no extra work was actually needed for it to show up. Magic! ✨ 

## Affects

Onboarding and making a donation using the PayPal Donations gateway.

### Account using PPCP
image

### Account using Express
image

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### Express
1. Go to the GiveWP General settings and set the country to Columbia
2. Go to the PayPal Donations connection screen
3. Connect using the following credentials:
    - Email: sb-siizv2863201@business.example.com
    - Pass: &;9jU+Y5
4. Make a donation

### PPCP
1. Go to the GiveWP General settings and set the country to United States
2. Go to the PayPal Donations connection screen
3. Connect using the following credentials:
    - Email: sb-kttpp2709027@business.example.com
    - Pass: 3lU<#qG.
4. Make a donation

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@ravinderk This is pretty well wrapped up and ready for review. My only question for you is how you'd recommend hiding/showing the "Donate Now" button. It's not used at all in Express, but we only want to remove it if it's the selected gateway. I'm sure this has been done with other gateways, so I figured I'd ask.

@DevinWalker Do you have any thoughts as far as design is concerned in Express?

**Comment by ravinderk** [comment]
@JasonTheAdams Testing credentials working fine for me.

File: src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php [hunk 1]
```diff
@@ -96,10 +97,11 @@ private function savePayPalMerchantDetails() {
...
 		$this->didWeGetValidSellerRestApiCredentials( $restApiCredentials );
 
-		$payPalAccount['clientId']       = $restApiCredentials['client_id'];
-		$payPalAccount['clientSecret']   = $restApiCredentials['client_secret'];
-		$payPalAccount['token']          = $tokenInfo;
-		$payPalAccount['accountIsReady'] = true;
+		$payPalAccount['clientId']               = $restApiCredentials['client_id'];
+		$payPalAccount['clientSecret']           = $restApiCredentials['client_secret'];
+		$payPalAccount['token']                  = $tokenInfo;
+		$payPalAccount['supportsCustomPayments'] = 'PPCP' === $partnerLinkInfo['product'];
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php:103]: I would've liked to pass a boolean to the front-end, but the dumb `wp_localize_script` function converts everything to strings.

I'm not sure how that relates to this line, though. I'm just storing the boolean value of whether or not the product is PPCP. 😕

File: src/PaymentGateways/PayPalCommerce/AjaxRequestHandler.php [hunk 2]
```diff
@@ -198,13 +198,16 @@ public function createOrder() {
...
-			'donationAmount' => $postData['give-amount'],
-			'payer'          => [
+			'formId'              => $formId,
+			'donationAmount'      => $postData['give-amount'],
+			'payer'               => [
 				'firstName' => $postData['give_first'],
 				'lastName'  => $postData['give_last'],
 				'email'     => $postData['give_email'],
 			],
+			'application_context' => [
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/PaymentGateways/PayPalCommerce/AjaxRequestHandler.php:208]: I'm sorry, I'm not understanding what you're saying, here. Don't we want to inform PayPal at this point to redact the need for shipping information when creating the order?

File: src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php [hunk 3]
```diff
@@ -279,7 +281,7 @@ private function refreshAccountStatus() {
 		/** @var MerchantDetail $merchantDetails */
 		$merchantDetails = give( MerchantDetail::class );
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php]: Using DI, now: 38b8d293

File: src/PaymentGateways/PayPalCommerce/AjaxRequestHandler.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/AjaxRequestHandler.php:208]: @JasonTheAdams Check red arrow in myscreenshot. I said that this is already defined in repository function.

File: src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php:103]: @JasonTheAdams Got it

File: src/PaymentGateways/PayPalCommerce/AjaxRequestHandler.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/PaymentGateways/PayPalCommerce/AjaxRequestHandler.php:219]: I'm just not understanding your point. This _is_ the repository function. Are you saying it's defined elsewhere, too?

**Comment by JasonTheAdams** [comment]
@ravinderk I added the country setting we discussed, but we're unable to save it until we get the UI corrected. Please pull this and do your thing.

**Comment by ravinderk** [comment]
> @ravinderk I added the country setting we discussed, but we're unable to save it until we get the UI corrected. Please pull this and do your thing.

Alright, I will ping you after I complete it.

**Comment by JasonTheAdams** [comment]
I'm going to merge this PR as it's fulfilled it's original intent. We may decide to make changes related to this, but it can be in another PR. I'd like this merged so Expess can be tested in #5079.

**Pull request merged**